### ISSUE TITLE
Inheritance of type var bounds not enforced

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#8400]
### Discussed in https://github.com/microsoft/pyright/discussions/8399

<div type='discussions-op-text'>

<sup>Originally posted by **BaconPancakes** July 12, 2024</sup>
If a type variable of a parent generic is bound to a type, shouldn't derived classes of that type be subject to the same bounds?

Consider the following behavior:

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoASAhgM4AWRARgDYCmAsAFBNMDG1pJUAQqbQOK0UtEElYBtACoAuKAApi5KnQCUAXWlMo2qADp9LRu048wAVxQATXiQFCRYuTbvDRE6khIw1K2QGJYRFooEXAQKFIQgA8EWlYYWktZDy9MLhQwfAolGgZGHT0DZiMOEi4AEQcAN0TBVzEpNSc%2BOocJSR9-QNiQkDCoDPwEEFpbFHxaGpRYMnM0Mh4W%2BzdGtKhKcytYMChsilytHX1dQ2My0wtLStEay1aVlO8u4bBciCIYMXW4ojNbRds9zEAHIuPAelUiOFhIkwTtKME9so8gVjkwgA)

```python
from typing import Hashable


class BaseGeneric[T: (Hashable)]:
    ...

class BoundBaseGeneric(BaseGeneric[list]): # type error as expected: list is not hashable
    ...

class DerivedGeneric[T](BaseGeneric[T]): # type error not present even though BaseGeneric[T] is bound to hashable
    ...

class BoundDerivedGeneric[list]: # problematic because BaseGeneric's type var needs to be hashable
    ...

```
</div>

**Comment by erictraut** [comment]
Note that there's a minor bug in the code sample. The `BoundDerivedGeneric` isn't subclassing from `BaseGeneric[list]`. It's defining a type variable called `list`. This doesn't affect the analysis, however.

This is also unrelated to the PEP 695 syntax. The same false negative exists when using the old syntax. And interestingly, mypy also has this same bug.

```python
from typing import Generic, Hashable, TypeVar

T = TypeVar("T", bound=Hashable)
S = TypeVar("S")

class BaseGeneric(Generic[T]):
    ...

class DerivedGeneric(BaseGeneric[S]):
    ...

class BoundDerivedGeneric(DerivedGeneric[list]):
    ...
```

**Comment by erictraut** [comment]
Ah, digging into this further, this isn't actually a bug. The problem is that the `__hash__` protocol in Python doesn't honor the Liskov Substitution Principle. If you use any type other than `Hashable` here, pyright and mypy both detect an error.

```python
from typing import Generic, TypeVar

T = TypeVar("T", bound=int)
S = TypeVar("S")

class BaseGeneric(Generic[T]):
    ...

class DerivedGeneric(BaseGeneric[S]):
    ...
```

That's because the upper bound of `S` is (implicitly) `object`, and `object` isn't assignable to the upper bound for `T` (`int`).

The problem with `Hashable` is that `object` is compatible with the `Hashable` protocol while some subclasses of `object` are not. 

I don't think there's much a type checker can do about this. It's a fundamental flaw in the way that Python defines hashability, and it's unlikely to ever be changed.

Closing since this isn't a bug in pyright.
